Commit-status-and-panel

abhinandan13jan opened the pull request:
## Fixes 
<!-- For e.g Fixes: https://issues.redhat.com/browse/HAC-XXX -->
https://issues.redhat.com/browse/HACBS-1136

## Description
<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->
Adds status and task side panel to commit details

## Type of change
<!-- Please delete options that are not relevant. -->

- [x] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Screen shots / Gifs for design review 
<!-- If change affects UI in any way, tag relevant UX people and add screenshots/gifs  -->

<img width="1390" alt="Screenshot 2022-10-10 at 6 33 39 PM" src="https://user-images.githubusercontent.com/24852534/194877379-90ded86e-77c9-4aa9-8162-5ecb6ee6c678.png">

<img width="1383" alt="Screenshot 2022-10-10 at 6 33 30 PM" src="https://user-images.githubusercontent.com/24852534/194877413-33da9a04-ab04-4cc4-9ebe-1591a8466a3d.png">

<img width="1404" alt="Screenshot 2022-10-14 at 7 27 47 PM" src="https://user-images.githubusercontent.com/24852534/195867739-847a1334-b919-4ed5-95d3-633759427c4f.png">
<img width="1388" alt="Screenshot 2022-10-14 at 7 28 10 PM" src="https://user-images.githubusercontent.com/24852534/195867749-40fcf43e-c87c-4845-b801-f25f77799e87.png">



## How to test or reproduce?
<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

<!-- - [ ] Test A -->
<!-- - [ ] Test B -->

<!-- **Test Configuration(s)**: -->
Added unit tests for commit-utils, CommitSidePanelHeader, CommitSidePanelbody


## Browser conformance: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge


<!-- ## Checklist: -->

<!-- 
- [ ] Code follows the style guidelines
- [ ] Self-reviewed the code
- [ ] Added comments in hard-to-understand areas
- [ ] Made corresponding changes to the documentation
- [ ] Changes generate no new warnings
- [ ] Added tests that prove this fix is effective or that the feature works
- [ ] New and existing unit tests pass locally with new changes
- [ ] Any dependent changes have been merged and published in downstream modules 
-->

Ranelim commented:
@abhinandan13jan looks very cool so far!

Few comments:

- Let's move the GitHub icon and the status label after the page title. I think @christianvogt mentioned it is doable now
- Did we include a "read more" after a long commit comment? I think we set up a max character of 72 and "read more" will just display the rest.
- On the side panel:
- I'm not sure what is "go-3-on-push..." is that a build pipeline placeholder?
- We should have a side panel title, displaying the full name of the node
- Below that, we should have an icon and generic name of the node type (in grey, body weight)
- For the status below, we should start that string with the corresponding icon (green/red/yellow/blue)

Thank you!

abhinandan13jan commented:
> @abhinandan13jan looks very cool so far!
> 
> Few comments:
> 
> * Let's move the GitHub icon and the status label after the page title. I think @christianvogt mentioned it is doable now
> * Did we include a "read more" after a long commit comment? I think we set up a max character of 72 and "read more" will just display the rest.
> * On the side panel:
> * I'm not sure what is "go-3-on-push..." is that a build pipeline placeholder?
> * We should have a side panel title, displaying the full name of the node
> * Below that, we should have an icon and generic name of the node type (in grey, body weight)
> * For the status below, we should start that string with the corresponding icon (green/red/yellow/blue)
> 
> Thank you!

1. Moved the GH icon to header
2. Yes, I updated to the read more, but this string is only 56 char long so it doesn't show. However, for demo I've set the read more length to 45 in screenshot
3. Go-random-number is the actual kubernetes object name for the pipeline. I've updated to use logical name "build/Release/Integration Test" , `Buld` in this case instead. We currently don't have the logical mapping/definitions in place with the `real object`(kubernetes PipelineRun CR) vs the `display texts`, which will be taken care of once @karthikjeeyar data model for the workflow visualization is in

karthikjeeyar commented on src/hacbs/components/Commits/CommitDetailsView.tsx:
```diff
@@ -106,112 +157,151 @@ const CommitDetailsView: React.FC<CommitDetailsViewProps> = ({ commitName, appli
+                    variant="plain"
+                    onClick={onStatusClick}
+                  >
+                    <StatusIconWithText status={commitStatus} />
```
This component is rendering [PF badge](https://patternfly-react.surge.sh/components/badge) component, but the designs has[ PF Label ](https://patternfly-react.surge.sh/components/label)component with background colors matching the status.

current: 
<img width="449" alt="image" src="https://user-images.githubusercontent.com/9964343/192455381-4d85fdcc-411c-468b-8d17-80122c47dd96.png">

designs:
<img width="521" alt="image" src="https://user-images.githubusercontent.com/9964343/192455433-ac4e2636-d791-40d4-94c9-6f8350ba259f.png">

karthikjeeyar commented on src/hacbs/components/Commits/CommitSidePanel.tsx:
```diff
@@ -0,0 +1,170 @@
+  setSelectedPipelineRun,
+  sortedPLRList,
+}) => {
+  const [currentStatus, setCurrentStatus] = React.useState<string>(initialStatus);
```
In this case, If you are expecting to sync your props (initialStatus) with useState variable then this will fail to sync. UseState will not re-initialise when the props change. You need to sync it either with useEffect or re-render this component with `key`.

karthikjeeyar replied:
Also the fact that this status will change on progression of events, better to call something other than `initialStatus`, maybe `currentStatus` ?

abhinandan13jan replied:
Renamed to commitStatus and moved inside useEffect

karthikjeeyar reviewed (commented):
<img width="472" alt="Screenshot 2022-09-27 at 12 41 53 PM" src="https://user-images.githubusercontent.com/9964343/192463664-de8261cf-52b6-4ce7-b054-14c495ae978c.png">

Choosing a different status in dropdown breaks the UI.

abhinandan13jan commented:
Updated PR
- Modularised CommitSidePanel into Two new components CommitSidePanelHeader and CommitSidePanelBody
- Updated `StatusIconWithText` to use `Label` instead of `Badge`
- Added test suites for CommitSidePanelHeader, CommitSidePanelBody & commit-utils
- Moved initialStatus as commitStatus inside useEffect

abhinandan13jan commented:
<img width="764" alt="Screenshot 2022-09-29 at 11 53 05 PM" src="https://user-images.githubusercontent.com/24852534/193112652-51bf42a3-5139-4262-a26d-360546ab2c62.png">

karthikjeeyar commented on src/shared/components/pipeline-run-logs/StatusIcon.tsx:
```diff
@@ -66,17 +66,26 @@ export const StatusIconWithText: React.FC<StatusIconProps & { text?: string }> =
 }) => {
   return (
-    <Badge isRead>
+    <Label>
```
Label here is missing color property, so the background is still grey.

![image](https://user-images.githubusercontent.com/9964343/193259511-3e2423d8-2a2f-4cea-9e05-c9b6e890d29a.png)

karthikjeeyar commented on src/shared/components/pipeline-run-logs/StatusIcon.tsx:
```diff
@@ -66,17 +66,26 @@ export const StatusIconWithText: React.FC<StatusIconProps & { text?: string }> =
-      {text ?? status}
-    </Badge>
+      </span>
+      <span
+        style={{
+          color: status
+            ? getRunStatusColor(status).pftoken.value
+            : getRunStatusColor(runStatus.Cancelled).pftoken.value,
+        }}
+      >
```
Refer my previous comment.  Once color prop is added on the label, you no longer need to set the color on icon, text seperately.

abhinandan13jan replied:
updated
<img width="1383" alt="Screenshot 2022-10-10 at 6 33 30 PM" src="https://user-images.githubusercontent.com/24852534/194877189-d3ee0fa5-f3be-459a-8df6-081cef864030.png">

abhinandan13jan commented:
/retest

Katka92 commented:
/retest

christianvogt commented on src/hacbs/components/Commits/CommitSidePanelHeader.tsx:
```diff
@@ -0,0 +1,118 @@
+          isOpen={isOpen}
+          dropdownItems={dropdownItems}
+        />
+        {sortedPLRList[currentStatus].length > 0 && (
+          <span className="pf-u-ml-xs pf-u-mr-xs" data-test="plr-count">
+            {`${currentIndex + 1}/${sortedPLRList[currentStatus].length}`}
+          </span>
+        )}
+        <Button
+          variant={ButtonVariant.control}
+          className="pf-u-p-sm"
+          isDisabled={currentIndex >= sortedPLRList[currentStatus].length - 1}
+          onClick={nextTask}
+          data-test="next-plr-button"
+        >
+          <AngleDownIcon />
+        </Button>
+        <Button
+          variant={ButtonVariant.control}
+          className="pf-u-p-sm"
+          isDisabled={currentIndex < 1}
+          onClick={previousTask}
+          data-test="previous-plr-button"
+        >
+          <AngleUpIcon />
+        </Button>
```
The spacing on these buttons don't look like the UX doc.
This is something we can also address afterwards.

But it seems like UX wants what looks like input group attachments. I'm not sure exactly though.
@Ranelim could you help identify the PF components you've used in this mock:
![image](https://user-images.githubusercontent.com/14068621/195392174-1c1dcb31-ee8f-406d-959a-5b9895876bbb.png)

christianvogt replied:
icon buttons don't have grey outlines. This is what confused me. If we don't need to have those grey outlines then we're good. I thought it was some other component / styling i wasn't aware of.

christianvogt commented on src/hacbs/utils/commits-utils.ts:
```diff
@@ -77,3 +86,38 @@ export const getCommitsFromPLRs = (plrList: PipelineRunKind[], limit?: number):
+    return '-';
+  }
+  const runType = plr?.metadata.labels[PipelineRunLabel.COMMIT_TYPE_LABEL];
+  if (!runType) {
+    return plr.metadata.name;
```
This seems like an incorrect value for when there is no type available.

abhinandan13jan commented on src/hacbs/components/Commits/CommitSidePanel.tsx:
```diff
@@ -0,0 +1,87 @@
+  const [currentStatus, setCurrentStatus] = React.useState<string>(commitStatus);
+  const [currentIndex, setCurrentIndex] = React.useState<number>(0);
+
+  React.useEffect(() => {
+    setCurrentStatus(commitStatus);
+  }, [commitStatus]);
```
For a pipeline in progress, the commitstatus would change. So we not capture that?

christianvogt replied:
But doesn't this cause the view to switch on the user while they may be viewing a different one?

abhinandan13jan commented on src/hacbs/components/Commits/CommitDetailsView.tsx:
```diff
@@ -71,19 +92,51 @@ const CommitDetailsView: React.FC<CommitDetailsViewProps> = ({ commitName, appli
+      // sort plr into respective status array
+      const plrStatus = pipelineRunFilterReducer(plr);
+      // put task in correct status array
+      statuses.map((status) => {
+        if (plrStatus === status) {
+          runs[status].push(plr);
+        }
+      });
```
No, the `statuses` array holds the order specified by UX for different `runStatuses`

christianvogt replied:
Sorry I corrected my suggestion. Not sure why i wrote `runStatus` while i meant `run`. 
See now.

The suggestion is to simply look at the current `run` status map and if it contains a value, we push to that array rather than having to perform a loop to find a match with `statuses` since `run` should have already be initialized to only contain entries that match those in `statuses`.

abhinandan13jan commented on src/hacbs/components/Commits/CommitSidePanel.tsx:
```diff
@@ -0,0 +1,87 @@
+  }, [commitStatus]);
+
+  const changeStatus = (status) => {
+    status !== currentStatus && setCurrentStatus(status);
+    currentIndex !== 0 && setCurrentIndex(0);
```
When user clicks on nextTask or PreviousTask, current index changes

abhinandan13jan commented:
updated @christianvogt @karthikjeeyar 
<img width="1404" alt="Screenshot 2022-10-14 at 7 27 47 PM" src="https://user-images.githubusercontent.com/24852534/195867590-4d20f810-142e-4068-873b-8a2a198ca5cd.png">
<img width="1388" alt="Screenshot 2022-10-14 at 7 28 10 PM" src="https://user-images.githubusercontent.com/24852534/195867610-c66c7275-5025-413e-b69b-515b66945820.png">

karthikjeeyar commented on src/hacbs/components/Commits/CommitDetailsView.tsx:
```diff
@@ -49,6 +53,23 @@ const CommitDetailsView: React.FC<CommitDetailsViewProps> = ({ commitName, appli
+  const [selectedPipelineRun, setSelectedPipelineRun] = React.useState<PipelineRunKind>(null);
+  const drawerRef = React.useRef<HTMLDivElement>();
+
+  // const onPanelExpand = () => {
+  //   drawerRef.current && drawerRef.current.focus();
+  // };
```
nit: remove commented code

karthikjeeyar commented on src/hacbs/components/Commits/CommitDetailsView.tsx:
```diff
@@ -71,19 +92,47 @@ const CommitDetailsView: React.FC<CommitDetailsViewProps> = ({ commitName, appli
+      // sort plr into respective status array
+      const plrStatus = pipelineRunFilterReducer(plr);
+      // put task in correct status array
+      statuses.forEach((status) => {
+        if (plrStatus === status) {
+          if (!runs[status]) {
+            runs[status] = [];
+          }
+          runs[status].push(plr);
+        }
+      });
+    });
```
you missed to change the suggestion from christian to remove the status map and directly push the `plrStatus` in runs array.

karthikjeeyar reviewed (commented):
Works as expected, still one more comment from christian needs to be resolved.

karthikjeeyar commented:
/lgtm

christianvogt commented:
/approve

karthikjeeyar commented:
/retest
